src/k2/dto/SKVRecord.cpp

```diff
@@ -93,15 +97,69 @@ Payload SKVRecord::getSharedPayload() {
     return storage.fieldData.shareAll();
 }
 
+// The constructor for an SKVRecord that a user of the SKV client would use to create a request
 SKVRecord::SKVRecord(const String& collection, std::shared_ptr<Schema> s) :
-            schema(s), collectionName(collection) {
+            schema(s), collectionName(collection), keyValuesAvailable(true), keyStringsConstructed(true) {
     storage.schemaVersion = schema->version;
     storage.fieldData = Payload(Payload::DefaultAllocator);
     partitionKeys.resize(schema->partitionKeyFields.size());
     rangeKeys.resize(schema->rangeKeyFields.size());
 }
 
-String SKVRecord::getPartitionKey() const {
+// The constructor for an SKVRecord that is created by the SKV client to be returned to the user in a response
+SKVRecord::SKVRecord(const String& collection, std::shared_ptr<Schema> s, Storage&& storage, bool keyAvail) :
+        schema(s), collectionName(collection), storage(std::move(storage)), keyValuesAvailable(keyAvail),
+        keyStringsConstructed(false) {}
+
+template <typename T>
+void SKVRecord::makeKeyString(std::optional<T> value, const String& fieldName, int tmp) {
+    (void) tmp;
+    (void) fieldName;
+    for (size_t i = 0; i < schema->partitionKeyFields.size(); ++i) {
+        // Subtracting 1 from fieldCursor here and below because this function gets called
+        // after the deserialize function moves the fieldCursor
+        if (schema->partitionKeyFields[i] == fieldCursor-1) {
+            if (value.has_value()) {
+                partitionKeys[i] = FieldToKeyString<T>(*value);
+            } else {
+                partitionKeys[i] = schema->fields[fieldCursor-1].nullLast ? NullLastToKeyString() : NullFirstToKeyString();
+            }
+        }
+    }
+
+    for (size_t i = 0; i < schema->rangeKeyFields.size(); ++i) {
+         if (schema->rangeKeyFields[i] == fieldCursor-1) {
+            if (value.has_value()) {
+                rangeKeys[i] = FieldToKeyString<T>(*value);
+            } else {
+                rangeKeys[i] = schema->fields[fieldCursor-1].nullLast ? NullLastToKeyString() : NullFirstToKeyString();
+            }
+        }
+    }
+}
+
+void SKVRecord::constructKeyStrings() {
+    partitionKeys.resize(schema->partitionKeyFields.size());
+    rangeKeys.resize(schema->rangeKeyFields.size());
+
+    uint32_t beginCursor = getFieldCursor();
+    seekField(0);
+    for (size_t i = 0; i < partitionKeys.size() + rangeKeys.size(); ++i) {
```

Comment:
Are we preventing the same field from being used in both partition and range key? i.e. is it possible for pk.size() + rk.size() to be < allFields.size()